(TSC 3.5.1) TS2590: Expression produces a union type that is too complex to represent.

<!--  
Thanks for contacting us! We're here to help.

Before you report an issue, check if it's been reported before:

  * Search: https://github.com/OfficeDev/office-ui-fabric-react/search?type=Issues
  * Search by area or component: https://github.com/OfficeDev/office-ui-fabric-react/issues/labels
  
Please provide a reproduction of the bug in a codepen, if possible. Here’s how:

  * Goto https://aka.ms/fabricpen for a starting codepen 
  * You can also use the "Export to Codepen" feature for the various components in our documentation site.
  * Alternatively, you can also use https://aka.ms/fabricdemo to get permanent repro links if the repro occurs with an example. 
    (A permanent link is preferable to "use the website" as the website can change)
 
Note that if you do not provide enough information to reproduce the issue, we may not be able to take action on your report.
-->

### Environment Information 
 
- __Package version(s)__: 6.182.1
- __Browser and OS versions__: Edge Dev 76.0.167.1

### Please provide a reproduction of the bug in a codepen:
Test Repo: https://github.com/chenxinyanc/fabric-test

In the root folder, run
```powershell
npm install
npm run build
```
<!-- Goto https://aka.ms/fabricpen for a starting codepen -->
<!-- See http://codepen.io/dzearing/pens/public/?grid_type=list for a variety of examples -->
<!-- Alternatively, you can also use https://aka.ms/fabricdemo to get permanent repro links. -->
 
#### Actual behavior:
When building with typescript 3.5.1, I see the following error
```
ERROR in [at-loader] ./node_modules/@uifabric/merge-styles/lib/IStyleSet.d.ts:49:75 
    TS2590: Expression produces a union type that is too complex to represent.

ERROR in [at-loader] ./node_modules/office-ui-fabric-react/lib/components/DetailsList/ShimmeredDetailsList.types.d.ts:10:18 
    TS2430: Interface 'IShimmeredDetailsListProps' incorrectly extends interface 'IDetailsListProps'.
  Types of property 'styles' are incompatible.
    Type 'IStyleFunctionOrObject<IShimmeredDetailsListStyleProps, IShimmeredDetailsListStyles>' is not assignable to type 'IStyleFunctionOrObject<IDetailsListStyleProps, IDetailsListStyles>'.    
      Type 'IStyleFunction<IShimmeredDetailsListStyleProps, IShimmeredDetailsListStyles>' is not assignable to type 'IStyleFunctionOrObject<IDetailsListStyleProps, IDetailsListStyles>'.
        Type 'IStyleFunction<IShimmeredDetailsListStyleProps, IShimmeredDetailsListStyles>' is not assignable to type 'IStyleFunction<IDetailsListStyleProps, IDetailsListStyles>'.
          Type 'IShimmeredDetailsListStyles' is missing the following properties from type 'IDetailsListStyles': focusZone, headerWrapper, contentWrapper
```
The file context with error is `TStyleSet['subComponentStyles'][P]` as in:
https://github.com/OfficeDev/office-ui-fabric-react/blob/ad9193054e3542ca04f265821fca34c51e68b9ff/packages/merge-styles/src/IStyleSet.ts#L35-L41
And the suffx `[P]` is the direct (but not root) cause of TS2590 error.
 
#### Expected behavior:
There shouldn't be any error, just like it when I'm building with typescript 3.4.5.

#### Priorities and help requested:
 
Are you willing to submit a PR to fix? Yes, but I'm currently not sure how to refactor [`IStyleSet`](https://github.com/OfficeDev/office-ui-fabric-react/blob/ad9193054e3542ca04f265821fca34c51e68b9ff/packages/merge-styles/src/IStyleSet.ts#L22) safely.
 
Requested priority: Normal
 
Products/sites affected: N/A

Here is the exact error I'm seeing, but I'm suspicious that it is coming from including a different dependency, since this isn't the first flaky "styles" related issue I've seen show up today.

```
../../node_modules/office-ui-fabric-react/lib/components/DetailsList/ShimmeredDetailsList.types.d.ts(10,18): error TS2430: Interface 'IShimmeredDetailsListProps' incorrectly extends interface 'IDetailsListProps'.
  Types of property 'styles' are incompatible.
    Type 'IStyleFunction<IShimmeredDetailsListStyleProps, IShimmeredDetailsListStyles> | Partial<IShimmeredDetailsListStyles> | undefined' is not assignable to type 'IStyleFunction<IDetailsListStyleProps, IDetailsListStyles> | Partial<IDetailsListStyles> | undefined'.
      Type 'IStyleFunction<IShimmeredDetailsListStyleProps, IShimmeredDetailsListStyles>' is not assignable to type 'IStyleFunction<IDetailsListStyleProps, IDetailsListStyles> | Partial<IDetailsListStyles> | undefined'.
        Type 'IStyleFunction<IShimmeredDetailsListStyleProps, IShimmeredDetailsListStyles>' is not assignable to type 'IStyleFunction<IDetailsListStyleProps, IDetailsListStyles>'.
          Type 'IShimmeredDetailsListStyles' is missing the following properties from type 'IDetailsListStyles': focusZone, headerWrapper, contentWrapper
npm ERR! code ELIFECYCLE
npm ERR! errno 2
```

--- a/common/changes/@uifabric/merge-styles/jg-fix-ts35-issues_2019-06-04-15-55.json
+++ b/common/changes/@uifabric/merge-styles/jg-fix-ts35-issues_2019-06-04-15-55.json
@@ -1,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "@uifabric/merge-styles",
+      "comment": "Fix TS 3.5 typing issues.",
+      "type": "minor"
+    }
+  ],
+  "packageName": "@uifabric/merge-styles",
+  "email": "jagore@microsoft.com"
+}

--- a/common/changes/office-ui-fabric-react/jg-fix-ts35-issues_2019-06-04-15-55.json
+++ b/common/changes/office-ui-fabric-react/jg-fix-ts35-issues_2019-06-04-15-55.json
@@ -1,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "office-ui-fabric-react",
+      "comment": "Fix TS 3.5 typing issues.",
+      "type": "minor"
+    }
+  ],
+  "packageName": "office-ui-fabric-react",
+  "email": "jagore@microsoft.com"
+}

--- a/packages/merge-styles/etc/merge-styles.api.md
+++ b/packages/merge-styles/etc/merge-styles.api.md
@@ -73,7 +73,7 @@ export type IProcessedStyleSet<TStyleSet extends IStyleSet<TStyleSet>> = {
     [P in keyof Omit<TStyleSet, 'subComponentStyles'>]: string;
 } & {
     subComponentStyles: {
-        [P in keyof TStyleSet['subComponentStyles']]: __MapToFunctionType<TStyleSet['subComponentStyles'][P]>;
+        [P in keyof TStyleSet['subComponentStyles']]: __MapToFunctionType<TStyleSet['subComponentStyles'] extends infer J ? (P extends keyof J ? J[P] : never) : never>;
     };
 };
 

--- a/packages/merge-styles/src/IStyleSet.ts
+++ b/packages/merge-styles/src/IStyleSet.ts
@@ -37,5 +37,9 @@ export type IConcatenatedStyleSet<TStyleSet extends IStyleSet<TStyleSet>> = {
  * into a class name. Additionally, all subComponentStyles are style functions.
  */
 export type IProcessedStyleSet<TStyleSet extends IStyleSet<TStyleSet>> = { [P in keyof Omit<TStyleSet, 'subComponentStyles'>]: string } & {
-  subComponentStyles: { [P in keyof TStyleSet['subComponentStyles']]: __MapToFunctionType<TStyleSet['subComponentStyles'][P]> };
+  subComponentStyles: {
+    [P in keyof TStyleSet['subComponentStyles']]: __MapToFunctionType<
+      TStyleSet['subComponentStyles'] extends infer J ? (P extends keyof J ? J[P] : never) : never
+    >
+  };
 };

--- a/packages/office-ui-fabric-react/etc/office-ui-fabric-react.api.md
+++ b/packages/office-ui-fabric-react/etc/office-ui-fabric-react.api.md
@@ -29,6 +29,7 @@ import { IStyleFunctionOrObject } from '@uifabric/utilities';
 import { IStyleSet } from '@uifabric/styling';
 import { ITheme } from '@uifabric/styling';
 import { KeyCodes } from '@uifabric/utilities';
+import { Omit } from '@uifabric/utilities';
 import * as PropTypes from 'prop-types';
 import * as React from 'react';
 
@@ -6680,7 +6681,7 @@ export interface IShimmerColors {
 }
 
 // @public
-export interface IShimmeredDetailsListProps extends IDetailsListProps {
+export interface IShimmeredDetailsListProps extends Omit<IDetailsListProps, 'styles'> {
     // @deprecated
     detailsListStyles?: IDetailsListProps['styles'];
     enableShimmer?: boolean;

--- a/packages/office-ui-fabric-react/src/components/DetailsList/ShimmeredDetailsList.types.ts
+++ b/packages/office-ui-fabric-react/src/components/DetailsList/ShimmeredDetailsList.types.ts
@@ -1,13 +1,13 @@
 import { IDetailsListProps } from './DetailsList.types';
 import { IDetailsRowProps } from './DetailsRow.types';
 import { IStyle } from '../../Styling';
-import { IStyleFunctionOrObject } from '../../Utilities';
+import { IStyleFunctionOrObject, Omit } from '../../Utilities';
 
 /**
  * ShimmeredDetailsList props interface
  * {@docCategory DetailsList}
  */
-export interface IShimmeredDetailsListProps extends IDetailsListProps {
+export interface IShimmeredDetailsListProps extends Omit<IDetailsListProps, 'styles'> {
   /**
    * DetailsList `styles` to pass through.
    * Due to ShimmeredDetailsList overriding the extended `styles` prop with its own one

